# Exploring eBay Car Sales Data

This is a project from DataQuest. Its goal is to clean and analyze a set of data about used car listings on the classifieds section of the German eBay. 

The data can no longer be found on Kaggle, but it can still be accessed [here](https://data.world/data-society/used-cars-data).

The dataset contains the following columns:

* `dateCrawled` - when the ad was first crawled; all field-values are taken from this date
* `name` - name of the car
* `seller` - whether the seller is private or a dealer
* `offerType` - the type of listing
* `price` - the price on the ad
* `abtest` - whether the listing is included in an A/B test
* `vehicleType` - the vehicle Type
* `yearOfRegistratio` - the year when the car was first registered
* `gearbox` - the transmission type
* `powerPS` - the power of the car in PS
* `model` - the car model name
* `kilometer` - how many kilometers the car has driven
* `monthOfRegistration` - the month when the car was first registered
* `fuelType` - what type of fuel the car uses
* `brand` - the brand of the car
* `notRepairedDamage` - if the car has a damage which is not yet repaired
* `dateCreated` - the date when the eBay listing was created
* `nrOfPictures` - the number of pictures in the ad
* `postalCode` - the postal code for the location of the vehicle
* `lastSeenOnline` - when the crawler saw this ad last online

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
# I will print the first few rows and some information
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Out of the 20 columns, 5 have null values; however, the column with the most null values is still less than 20% null. Additionally, 5 columns are integers, while the rest are strings.

# Cleaning the Columns

The column names use camelcase instead of snakecase, but we'll convert them to snakecase to make it easier to work with.

Additionally, we'll need to reword some column names to be more descriptive or easier to understand.

In [3]:
# I will print the names of the columns as is
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
# To make my changes, I will copy that array
# And then I will make edits manually
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

# I will print out some of the data to see my changes
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Cleaning the Data

In this section, we'll explore the data to determine any further cleaning that we will have to do.

We will first look for text columns where all or almost all values are the same; these can be dropped because they yield no useful information. And we will look for numeric data that is stored as text; this can be cleaned and converted.

In [5]:
# I will look at some descriptive statistics of the columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-23 18:39:34  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

From the data above, `seller` and `offer_type` have mostly the same value: "privat" and "Angebot", respectively; they can be dropped.

The `num_pictures` column seems a little odd and inconsistent. So, that will need some investigating.

Additionally, so will `registration_year`. Cars did not exist in the year 1000 and we have not yet reached the year 9999.

The `price` and `odometer` columns have numeric data stored as text; the have extra characters like "$" and "km", so they will need to be cleaned and converted.

In [6]:
autos["num_pictures"].value_counts()

0    50000
Name: num_pictures, dtype: int64

So, for all 50,000 entries, there are 0 photos in `num_pictures`; I will drop this column, along with `seller` and `offer_type`.

In [7]:
autos = autos.drop(["num_pictures", "seller", "offer_type"], axis=1)

Now, I will clean up the `price` and `odometer` columns and convert them to integers.

In [8]:
# I want to price out the values of the prices to known
# what characters to remove
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [9]:
# This is for the price column
autos["price"] = (autos["price"].str.replace("$","")
                 .str.replace(",","")
                 .astype(int)
                 )

# This is to get a preview of the converted data
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [10]:
# I want to price out the values of the odometer to known
# what characters to remove
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [11]:
# This is for the odometer column
autos["odometer"] = (autos["odometer"].str.replace("km","")
                 .str.replace(",","")
                 .astype(int)
                 )

# I will rename the odometer column to include the unit
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
# This is to get a preview of the converted data
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

# Exploring Odometer and Price

In this section, we will look at some unrealistic values in the two columns and remove those.

In [12]:
# I will explore the odometer values
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
autos["odometer_km"].value_counts()

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

It seems like there are more high-mileage listings than low-mileage listings.

Additionally, they have all been rounded to the nearest thousand. Seeing as there are only 13 unique price values, this might indicate that the sellers had to choose from preset options.

In [13]:
# I will explore the price values
print(autos["price"].unique().shape)
print(autos["price"].describe())

# Given the size of the price values, I will only
# print up to 20 of them
autos["price"].value_counts().head(20)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

The prices are also rounded, much like the odometer data. But, given the larger amount of unique values, it could be possible that the sellers had a tendency to round, rather than being constrained by preset options.

Over 1400 prices have a value of "0". These can be eliminated, especially since they account for less than 3% of the data in the `price` column.

The highest price is $100,000,000; now that is quite a large number, so that will need further investigation. Let's take a look at the higher values.

In [14]:
# I will look at the highest prices
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [15]:
# And I will look at the lowest prices
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are some listins where the price is 35 dollars or under. Any price of 35 dollars or less is low for a car. But, since eBay is an auction site, these prices do seem believable, as they might represent low bids.

On the other hand, one hundred million dollars is very high. The prices rise steadily until 350,000 dollars; after that, they jump dramatically. So, we could remove any data with a price over 350,000 dollars.

In [16]:
# I will remove 0 and prices over $350,000 with a Boolean index
autos[autos["price"].between(1,350001)]
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

# Exploring the Date Columns

There are five columns with dates, with two created by the crawler and the other three created by the website.

* `date_crawled` - added by the crawler
* `last_seen` - added by the crawler
* `ad_created` - from the website
* `registration_month` - from the website
* `registration_year` - from the website

Three columns, `date_crawled`, `last_seen`, and `ad_created`, are string values. We will have to convert these into integers to work with them.

In [17]:
# I will print out the three columns to see how they
# are formatted
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 characters represent the day. So, I will extract those to work with. After, I will generate a distribution and then sort by index.

In [18]:
(autos['date_crawled'].str[:10]
                      .value_counts(normalize=True, dropna=False)
                      .sort_index()
                      )

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

The data seems to have been crawled daily over a period of one month. Most of the listings seem equally distributed.

In [19]:
(autos['ad_created'].str[:10]
                      .value_counts(normalize=True, dropna=False)
                      .sort_index()
                      )

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

Most listings seem to have been created within 1 month of the listing date. Some go as far back as 9 months.

In [20]:
(autos['last_seen'].str[:10]
                      .value_counts(normalize=True, dropna=False)
                      .sort_index()
                      )

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

The data seems to have been crawled daily over a period of one month too.

This column shows the dates when the crawler last saw the listing. The last three days show a disproportionate amount of listings closing; they are 6-10 times the proportion of the rest.

This is likely the result of the listing period ending, rather than the cars being sold, as a disproportionate jump in sales in three days seems unlikely.

In [21]:
# I will see the distribution of the registration years
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As mentioned before, some registration years make no sense. Cars did not exist in the year 1000 and we have not yet reached the year 9999.

# Dealing with Incorrect Years

Along with those weird years that we have just seen, it would also be impossible for a car to be registered after 2017; the listing was first seen in 2016.

Cars registered within the first few decades of 1900 seem plausible.

So, we could remove listings outside the range of 1900-2016.

In [22]:
# I will check the out amount of listings outside the range
(~autos["registration_year"].between(1900,2016)).value_counts(normalize=True)

False    0.96056
True     0.03944
Name: registration_year, dtype: float64

As we can see from the result, the amount of values outside the range of 1900-2016, marked `True`, is less 4% of the data. So, we can remove them.

In [23]:
# With a Boolean index, I will remove the unwanted values
# and I will print out the values that fall within the range
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
2006    0.056384
2001    0.056280
2002    0.052740
1998    0.051074
2007    0.047972
Name: registration_year, dtype: float64

It seems that most cars were registered within the past 20 years.

# Exploring Price by Brand

Now, we will aggregate the data on car brands to explore variations across them.

In [28]:
# I will first find the frequencies of the brands
autos["brand"].value_counts(normalize=True)

volkswagen        0.212126
bmw               0.110019
opel              0.108166
mercedes_benz     0.095361
audi              0.086387
ford              0.069793
renault           0.047347
peugeot           0.029524
fiat              0.025860
seat              0.018177
skoda             0.016032
mazda             0.015137
nissan            0.015095
citroen           0.013929
smart             0.013909
toyota            0.012472
sonstige_autos    0.010952
hyundai           0.009848
volvo             0.009245
mini              0.008641
mitsubishi        0.008141
honda             0.007850
kia               0.007100
alfa_romeo        0.006621
porsche           0.006101
suzuki            0.005913
chevrolet         0.005705
chrysler          0.003665
daihatsu          0.002561
dacia             0.002561
jeep              0.002249
subaru            0.002186
land_rover        0.002040
saab              0.001603
jaguar            0.001582
trabant           0.001562
daewoo            0.001499
r

German manufacturers account for half of the listings. Volkswagen forms the largest brand, with almost twice the next largest listing.

Many brands account for less than 1% of the listings; any listings under 5% will be deleted.

In [29]:
# With Boolean indexing, I will filter out those under 5%
brand_counts = autos["brand"].value_counts(normalize=True)
larger_brands = brand_counts[brand_counts > 0.05].index
print(larger_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [30]:
# Now, I will aggregate the data
brand_avg_prices = {}

for brand in larger_brands:
    selected = autos[autos["brand"] == brand]
    mean = selected["price"].mean()
    brand_avg_prices[brand] = int(mean)
    
brand_avg_prices

{'volkswagen': 6516,
 'bmw': 8334,
 'opel': 5252,
 'mercedes_benz': 30317,
 'audi': 9093,
 'ford': 7263}

Of the brands that account for over 5%, we can see that:
* BMW, Mercedes Benz, and Audi are the most expensive
* Ford and Opel are the least expensive
* Volkswagen are in between

# Exploring Mileage by Brand

In this section, we will aggregate the mileage data.

We will then compare the price and mileage by brand by placing both aggregations into a dataframe and displaying them side-by-side.

In [36]:
# I will put the price aggregations into a dataframe
# I need to first convert the dictionary into a series
bap_series = pd.Series(brand_avg_prices)
mean_price = pd.DataFrame(bap_series, columns=['mean_price'])
mean_price

mean_price
volkswagen           6516
bmw                  8334
opel                 5252
mercedes_benz       30317
audi                 9093
ford                 7263

In [34]:
# Now, I will aggregate the data for mileage
brand_avg_miles = {}

for brand in larger_brands:
    selected = autos[autos["brand"] == brand]
    mean = selected["odometer_km"].mean()
    brand_avg_miles[brand] = int(mean)
    
brand_avg_miles

{'volkswagen': 128730,
 'bmw': 132434,
 'opel': 129227,
 'mercedes_benz': 130860,
 'audi': 129287,
 'ford': 124046}

In [44]:
# I will put the mileage data into a dataframe using
# the same process
# I will sort the mileages for easier readability
bam_series = pd.Series(brand_avg_miles).sort_values(ascending=False)
pd.DataFrame(bam_series, columns=['mean_mileage'])

mean_mileage
bmw                  132434
mercedes_benz        130860
audi                 129287
opel                 129227
volkswagen           128730
ford                 124046

In [46]:
# Finally, I will combine both dataframes for comparison
brand_info = pd.DataFrame(bam_series, columns=['mean_mileage'])
brand_info["mean_price"] = mean_price
brand_info

mean_mileage  mean_price
bmw                  132434        8334
mercedes_benz        130860       30317
audi                 129287        9093
opel                 129227        5252
volkswagen           128730        6516
ford                 124046        7263

It appears that the mileages do not vary by brand as much as the prices do. The more expensive cars seem to rack up more mileage, whereas the cheaper cars rack up less.

# Conclusion

There is a slight correlation of mileage with price; more expensive brands have higher mileages.

This project was mainly intended to demonstrate data cleaning techniques, so there is not as much analysis this time.